x/evm/keeper/msg_server.go

```diff
@@ -318,86 +318,62 @@ func (k *Keeper) ApplyEvmMsg(
-	evmObj.StateDB.SetNonce(sender.Address(), msg.Nonce()+1)
+	evmObj.StateDB.SetNonce(msg.From(), msg.Nonce()+1)
 
 	// EVM execution error needs to be available for the JSON-RPC client
 	var vmError string
 	if vmErr != nil {
 		vmError = vmErr.Error()
 	}
 
-	// The dirty states in `StateDB` is either committed or discarded after return
-	if commit {
-		if err := evmObj.StateDB.(*statedb.StateDB).Commit(); err != nil {
-			return nil, errors.Wrap(err, "ApplyEvmMsg: failed to commit stateDB")
-		}
-	}
-	// Rare case of uint64 gas overflow
-	if msg.Gas() < leftoverGas {
-		return nil, errors.Wrapf(core.ErrGasUintOverflow, "ApplyEvmMsg: message gas limit (%d) < leftover gas (%d)", msg.Gas(), leftoverGas)
-	}
+	// process gas refunds (we refund a portion of the unused gas)
+	gasUsed := msg.Gas() - gasRemaining
+	refundAmount := GasToRefund(evmObj.StateDB.GetRefund(), gasUsed, fullRefundLeftoverGas)
+	gasRemaining += refundAmount
+	gasUsed -= refundAmount
 
-	// TODO: UD-DEBUG: Clarify text below.
-	// GAS REFUND
-	// If msg.Gas() > gasUsed, we need to refund extra gas.
-	// leftoverGas = amount of extra (not used) gas.
-	// If the msg comes from user, we apply refundQuotient capping the refund to 20% of used gas
-	// If msg is internal (funtoken), we refund 100%
-	//
-	// EIP-3529: refunds are capped to gasUsed / 5
-	// We evaluate "fullRefundLeftoverGas" and use only the gas consumed (not the
-	// gas limit) if the `ApplyEvmMsg` call originated from a state transition
-	// where the chain set the gas limit and not an end-user.
-	refundQuotient := params.RefundQuotientEIP3529
-	if fullRefundLeftoverGas {
-		refundQuotient = 1 // 100% refund
+	evmResp := &evm.MsgEthereumTxResponse{
+		GasUsed: gasUsed,
+		VmError: vmError,
+		Ret:     returnBz,
+		Logs:    evm.NewLogsFromEth(evmObj.StateDB.(*statedb.StateDB).Logs()),
+		Hash:    txHash.Hex(),
 	}
-	temporaryGasUsed := msg.Gas() - leftoverGas
-	refund := GasToRefund(evmObj.StateDB.GetRefund(), temporaryGasUsed, refundQuotient)
 
-	// update leftoverGas and temporaryGasUsed with refund amount
-	leftoverGas += refund
-	temporaryGasUsed -= refund
-	if msg.Gas() < leftoverGas {
-		return nil, errors.Wrapf(core.ErrGasUintOverflow, "ApplyEvmMsg: message gas limit (%d) < leftover gas (%d)", msg.Gas(), leftoverGas)
+	if gasRemaining > msg.Gas() {
+		evmResp.GasUsed = msg.Gas() // cap the gas used to the original gas limit
+		return evmResp, errors.Wrapf(core.ErrGasUintOverflow, "ApplyEvmMsg: message gas limit (%d) < leftover gas (%d)", msg.Gas(), gasRemaining)
 	}
 
-	// Min gas used is a % of gasLimit
-	gasUsed := math.LegacyNewDec(int64(temporaryGasUsed)).TruncateInt().Uint64()
-
-	// This resulting "leftoverGas" is used by the tracer. This happens as a
-	// result of the defer statement near the beginning of the function with
-	// "vm.Tracer".
-	leftoverGas = msg.Gas() - gasUsed
+	// The dirty states in `StateDB` is either committed or discarded after return
+	if commit {
```

Comment:
I moved the StateDB.Commit() to after the gas calculations so we have accurate gas accounting prior to commit and return.